aio-libs/aiohttp-admin
# Dynamic component draft

To test:

- Clone this branch and `cd aiohttp-admin`.
- Python dependencies: `pip install -r requirements.txt`
- Build the JS module: `cd admin-js; yarn install; yarn build`
- Try the example:
  - `cd ../examples`
  - `PYTHONPATH='..' python simple.py`
  - Go to http://localhost:8080/admin
  - You can login with admin/admin and see the basic project working.

So, the library has the JS module in admin-js/ pre-compiled and automatically served for users. The entire admin interface is created from a JSON schema (dumped into the HTML: https://github.com/aio-libs/aiohttp-admin/blob/6f2c9a2a9dee246aa2506284c67c567fe6312ab5/aiohttp_admin/views.py#L23) and read automatically by the JS file loaded from the same HTML page.

-----

The goal is to then allow users to add functions and components dynamically at runtime without needing to rebuild the library's module itself. Adding functions is easy and you can test this with validators.py:
`PYTHONPATH='..' python validators.py`
Then in the admin, edit a user, and you'll see that the votes field will only allow odd numbers.

The validator function for odd numbers is defined in a tiny JS module (I've served it as a hardcoded endpoint to keep the example self-contained, but obviously this would be a static file normally): https://github.com/aio-libs/aiohttp-admin/blob/6f2c9a2a9dee246aa2506284c67c567fe6312ab5/examples/validators.py#L15-L21
The URL to that JS module is included into the JSON schema: https://github.com/aio-libs/aiohttp-admin/blob/6f2c9a2a9dee246aa2506284c67c567fe6312ab5/examples/validators.py#L78
The library's JS module then imports the given URL and adds the available functions/components to the available ones that the user can reference: https://github.com/aio-libs/aiohttp-admin/blob/6f2c9a2a9dee246aa2506284c67c567fe6312ab5/admin-js/src/App.js#L74-L81
And the function is just referenced by name in the JSON schema as well: https://github.com/aio-libs/aiohttp-admin/blob/6f2c9a2a9dee246aa2506284c67c567fe6312ab5/examples/validators.py#L76

-----

Doing this for React components is not easy. I can't figure out how to do this. It seems that because the components are compiled in different modules, they refer to different states and so a component from the user module can't be used in the library's module when imported at runtime.

More comments below.
